nomad/secure_variables_endpoint.go

```diff
@@ -1,133 +1,470 @@
+		// At this point, we do not have a valid ACL token, nor are we being
+		// called, or able to confirm via the state store, by a node.
+		if node == nil {
+			return structs.ErrTokenNotFound
+		}
+	}
 
 	return nil
 }
 
-func (sv *SecureVariables) Update(args *structs.SecureVariablesUpsertRequest, reply *structs.SecureVariablesUpsertResponse) error {
-	if done, err := sv.srv.forward("SecureVariables.Update", args, args, reply); done {
+// List is used to list secure variables held within state. It supports single
+// and wildcard namespace listings.
+func (sv *SecureVariables) List(
+	args *structs.SecureVariablesListRequest,
+	reply *structs.SecureVariablesListResponse) error {
+
+	if done, err := sv.srv.forward(structs.SecureVariablesListRPCMethod, args, args, reply); done {
 		return err
 	}
+	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "list"}, time.Now())
 
-	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "update"}, time.Now())
+	// If the caller has requested to list secure variables across all namespaces, use
+	// the custom function to perform this.
+	if args.RequestNamespace() == structs.AllNamespacesSentinel {
+		return sv.listAllSecureVariables(args, reply)
+	}
 
-	// TODO: implement real ACL checks
-	if aclObj, err := sv.srv.ResolveToken(args.AuthToken); err != nil {
+	// Perform our mixed auth handling.
+	if err := sv.handleMixedAuthEndpoint(args.QueryOptions, acl.NamespaceCapabilityReadJob); err != nil {
 		return err
-	} else if aclObj != nil && !aclObj.IsManagement() {
-		return structs.ErrPermissionDenied
 	}
 
-	// TODO: implementation
-	SV_Upsert(args, reply)
+	// Set up and return the blocking query.
+	return sv.srv.blockingRPC(&blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, stateStore *state.StateStore) error {
 
-	return nil
+			// Perform the state query to get an iterator.
+			iter, err := stateStore.GetSecureVariablesByNamespaceAndPrefix(ws, args.RequestNamespace(), args.Prefix)
+			if err != nil {
+				return err
+			}
+
+			// Generate the tokenizer to use for pagination using namespace and
+			// ID to ensure complete uniqueness.
+			tokenizer := paginator.NewStructsTokenizer(iter,
+				paginator.StructsTokenizerOptions{
+					WithNamespace: true,
+					WithID:        true,
+				},
+			)
+
+			// Set up our output after we have checked the error.
+			var svs []*structs.SecureVariableStub
+
+			// Build the paginator. This includes the function that is
+			// responsible for appending a registration to the secure variables
+			// array.
+			paginatorImpl, err := paginator.NewPaginator(iter, tokenizer, nil, args.QueryOptions,
+				func(raw interface{}) error {
+					sv := raw.(*structs.SecureVariable)
+					svStub := sv.Stub()
+					svs = append(svs, &svStub)
+					return nil
+				})
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "failed to create result paginator: %v", err)
```

Comment:
I don't know that we want to add the HTTP status codes in the RPC handler? But I think this (and the one below) are internal errors and not bad requests if we do.

Reply:
I agree.  I think that we should be coding it with internal server error, but this particular code is in all of the other list endpoints that have been paginated in this way.  Should we maintain consistency with a TODO to move them all to internal errors or start doing the new (more correct) thing here and PR the other endpoints to move them to Internal Server Errors?

Reply:
I took a second look at where the error is coming from: [`paginator.go#L37-L64`](https://github.com/hashicorp/nomad/blob/v1.3.0/nomad/state/paginator/paginator.go#L37-L64). The only way we can get an error from `NewPaginator` is if the filter expression is bad, which _is_ a Bad Request! The only way we can get an error from `Page()` is if the append func throws an error, which we never actually do. 

So I think what you've got here is actually fine.